### ISSUE TITLE
Libvirt [release-4.8]: SIGSEGV: segmentation violation on executing openshift-install binary commands

### ISSUE THREAD

**Issue by pratham-m** [openshift/installer#4893]
# Version
Using `release-4.8` branch.
```console
$ openshift-install version
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x0]

goroutine 1 [running]:
github.com/aws/aws-sdk-go/service/directoryservice.init()
        /root/ipi/installer/vendor/github.com/aws/aws-sdk-go/service/directoryservice/errors.go:210 +0x24
```

# Platform: libvirt IPI

# What happened?
The `openshift-install` binary builds but throws `SIGSEGV: segmentation violation` error when executing any command on `ppc64le`

# What you expected to happen?
`openshift-install` binary should work properly and should not give `SIGSEGV` error on `ppc64le`

# How to reproduce it (as minimally and precisely as possible)?

Steps used to build the binary:
```console
$ git clone https://github.com/openshift/installer.git
$ git checkout release-4.8
$ export GOFLAGS="-mod=vendor"
$ export GO111MODULE=auto
$ TAGS=libvirt hack/build.sh
$ bin/openshift-install version
```

# Anything else we need to know?

```
$ TAGS=libvirt hack/build.sh
+ minimum_go_version=1.14
++ go version
++ cut -d ' ' -f 3
+ current_go_version=go1.15.7
++ version 1.15.7
++ IFS=.
++ printf '%03d%03d%03d\n' 1 15 7
++ unset IFS
++ version 1.14
++ IFS=.
++ printf '%03d%03d%03d\n' 1 14
++ unset IFS
+ '[' 001015007 -lt 001014000 ']'
+ MODE=release
++ git rev-parse --verify 'HEAD^{commit}'
+ GIT_COMMIT=da0ed4e925c093ed6e049c11d2bc68d562cc8d54
++ git describe --always --abbrev=40 --dirty
+ GIT_TAG=unreleased-master-4526-gda0ed4e925c093ed6e049c11d2bc68d562cc8d54-dirty
+ GOFLAGS=-mod=vendor
+ LDFLAGS=' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-4526-gda0ed4e925c093ed6e049c11d2bc68d562cc8d54-dirty -X github.com/openshift/installer/pkg/version.Commit=da0ed4e925c093ed6e049c11d2bc68d562cc8d54'
+ TAGS=libvirt
+ OUTPUT=bin/openshift-install
+ export CGO_ENABLED=0
+ CGO_ENABLED=0
+ case "${MODE}" in
+ LDFLAGS=' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-4526-gda0ed4e925c093ed6e049c11d2bc68d562cc8d54-dirty -X github.com/openshift/installer/pkg/version.Commit=da0ed4e925c093ed6e049c11d2bc68d562cc8d54 -s -w'
+ TAGS='libvirt release'
+ test '' '!=' y
+ go generate ./data
writing assets_vfsdata.go
+ echo 'libvirt release'
+ grep -q libvirt
+ export CGO_ENABLED=1
+ CGO_ENABLED=1
+ case "$(go env GOARCH)" in
++ go env GOARCH
+ LDFLAGS=' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-4526-gda0ed4e925c093ed6e049c11d2bc68d562cc8d54-dirty -X github.com/openshift/installer/pkg/version.Commit=da0ed4e925c093ed6e049c11d2bc68d562cc8d54 -s -w -linkmode external'
+ export CGO_ENABLED=1
+ CGO_ENABLED=1
+ go build -mod=vendor -ldflags ' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-4526-gda0ed4e925c093ed6e049c11d2bc68d562cc8d54-dirty -X github.com/openshift/installer/pkg/version.Commit=da0ed4e925c093ed6e049c11d2bc68d562cc8d54 -s -w -linkmode external' -tags 'libvirt release' -o bin/openshift-install ./cmd/openshift-install
```
```
$ bin/openshift-install version
unexpected fault address 0x1000100010000
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x3 addr=0x1000100010000 pc=0x1000100010000]

goroutine 1 [running, locked to thread]:
runtime.throw(0x231bd1d4, 0x5)
        /usr/local/go/src/runtime/panic.go:1116 +0x5c fp=0xc000d9fca8 sp=0xc000d9fc68 pc=0x168ed39c
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:749 +0x3e4 fp=0xc000d9fce8 sp=0xc000d9fca8 pc=0x169041f4
github.com/aws/aws-sdk-go/service/directoryservice.init()
        /root/ipi/installer/vendor/github.com/aws/aws-sdk-go/service/directoryservice/errors.go:210 +0x24 fp=0xc000d9fd58 sp=0xc000d9fd08 pc=0x1a0fd164
runtime.doInit(0x2b5796e0)
        /usr/local/go/src/runtime/proc.go:5652 +0xac fp=0xc000d9fda0 sp=0xc000d9fd58 pc=0x168fd5bc
runtime.doInit(0x2b5cad80)
        /usr/local/go/src/runtime/proc.go:5647 +0x60 fp=0xc000d9fde8 sp=0xc000d9fda0 pc=0x168fd570
runtime.doInit(0x2b59a380)
        /usr/local/go/src/runtime/proc.go:5647 +0x60 fp=0xc000d9fe30 sp=0xc000d9fde8 pc=0x168fd570
runtime.doInit(0x2b58bd20)
        /usr/local/go/src/runtime/proc.go:5647 +0x60 fp=0xc000d9fe78 sp=0xc000d9fe30 pc=0x168fd570
runtime.doInit(0x2b5a58c0)
        /usr/local/go/src/runtime/proc.go:5647 +0x60 fp=0xc000d9fec0 sp=0xc000d9fe78 pc=0x168fd570
runtime.doInit(0x2b581f60)
        /usr/local/go/src/runtime/proc.go:5647 +0x60 fp=0xc000d9ff08 sp=0xc000d9fec0 pc=0x168fd570
runtime.doInit(0x2b5a90e0)
        /usr/local/go/src/runtime/proc.go:5647 +0x60 fp=0xc000d9ff50 sp=0xc000d9ff08 pc=0x168fd570
runtime.main()
        /usr/local/go/src/runtime/proc.go:191 +0x1c4 fp=0xc000d9ffc0 sp=0xc000d9ff50 pc=0x168efbc4
runtime.goexit()
        /usr/local/go/src/runtime/asm_ppc64x.s:884 +0x4 fp=0xc000d9ffc0 sp=0xc000d9ffc0 pc=0x16922a64

goroutine 19 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x2b72edc0)
        /root/ipi/installer/vendor/k8s.io/klog/v2/klog.go:1164 +0x78
created by k8s.io/klog/v2.init.0
        /root/ipi/installer/vendor/k8s.io/klog/v2/klog.go:418 +0xe4

goroutine 13 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc00000c100)
        /root/ipi/installer/vendor/go.opencensus.io/stats/view/worker.go:276 +0xd4
created by go.opencensus.io/stats/view.init.0
        /root/ipi/installer/vendor/go.opencensus.io/stats/view/worker.go:34 +0x78

goroutine 28 [chan receive]:
k8s.io/klog.(*loggingT).flushDaemon(0x2b72ece0)
        /root/ipi/installer/vendor/k8s.io/klog/klog.go:1010 +0x78
created by k8s.io/klog.init.0
        /root/ipi/installer/vendor/k8s.io/klog/klog.go:411 +0xe0

goroutine 29 [sleep]:
time.Sleep(0x6fc23ac00)
        /usr/local/go/src/runtime/time.go:188 +0xc4
sigs.k8s.io/controller-runtime/pkg/log.init.0.func1()
        /root/ipi/installer/vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:63 +0x3c
created by sigs.k8s.io/controller-runtime/pkg/log.init.0
        /root/ipi/installer/vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:62 +0x34
```

# References

**Comment by pratham-m** [comment]
Thanks for the info @wking

**Comment by pratham-m** [comment]
Hi @wking I tried the changes in #4890 which basically removes `-linkmode external` for `ppc64le`, but I am still getting same segmentation fault. Am I missing something here?

```
+ go generate ./data
writing assets_vfsdata.go
+ echo 'libvirt release'
+ grep -q libvirt
+ export CGO_ENABLED=1
+ CGO_ENABLED=1
++ go env GOARCH
+ test ppc64le = arm64
+ go build -mod=vendor -ldflags ' -X github.com/openshift/installer/pkg/version.Raw=unreleased-master-4526-gda0ed4e925c093ed6e049c11d2bc68d562cc8d54-dirty -X github.com/openshift/installer/pkg/version.Commit=da0ed4e925c093ed6e049c11d2bc68d562cc8d54 -s -w' -tags 'libvirt release' -o bin/openshift-install ./cmd/openshift-install

$ bin/openshift-install version
unexpected fault address 0xffffffffffff1658
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x3 addr=0xffffffffffff1658 pc=0x1bcb9fe8]

goroutine 1 [running, locked to thread]:
runtime.throw(0x231bd1d4, 0x5)
        /usr/local/go/src/runtime/panic.go:1116 +0x5c fp=0xc000d1fca8 sp=0xc000d1fc68 pc=0x168ed39c
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:749 +0x3e4 fp=0xc000d1fce8 sp=0xc000d1fca8 pc=0x169041f4
github.com/aws/aws-sdk-go/service/directoryservice.init()
```

**Comment by dbenoit17** [comment]
This will not be fixed by https://github.com/openshift/installer/pull/4890, because libvirt mode sets CGO_ENABLED=1 regardless of -linkmode external.

**Comment by dbenoit17** [comment]
This should be fixed by https://github.com/golang/go/commit/9ed736ac2a99aa2e7ef7d8bed3b01ca8b20a6f80.  We should consider keeping this issue open until the upstream commit lands in a stable release that we can recommend to use.

**Comment by pratham-m** [comment]
thanks for the fix and update @dbenoit17

**Comment by pratham-m** [comment]
Hi @dbenoit17 do you know if the fix was merged upstream to a stable release?

**Comment by dbenoit17** [comment]
Yes, this appears to be resolved in the stable versions of Go 1.15.13,  Go 1.16.5, and later.

**Comment by pratham-m** [comment]
@wking can you help in adding below info to appropriate document?

> 
> 
> Yes, this appears to be resolved in the stable versions of Go 1.15.13, Go 1.16.5, and later.

**Comment by patrickdillon** [comment]
Fixed by #5120 

/close
